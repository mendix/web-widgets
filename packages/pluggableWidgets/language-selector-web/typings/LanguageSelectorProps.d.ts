/**
 * This file was generated from LanguageSelector.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { CSSProperties } from "react";
import { DynamicValue, ListValue, ListExpressionValue } from "mendix";

export type PositionEnum = "left" | "right" | "top" | "bottom";

export type TriggerEnum = "click" | "hover";

export interface LanguageSelectorContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    languageOptions: ListValue;
    languageCaption: ListExpressionValue<string>;
    position: PositionEnum;
    trigger: TriggerEnum;
    hideForSingle: boolean;
    screenReaderLabelCaption?: DynamicValue<string>;
}

export interface LanguageSelectorPreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
    renderMode?: "design" | "xray" | "structure";
    languageOptions: {} | { caption: string } | { type: string } | null;
    languageCaption: string;
    position: PositionEnum;
    trigger: TriggerEnum;
    hideForSingle: boolean;
    screenReaderLabelCaption: string;
}
