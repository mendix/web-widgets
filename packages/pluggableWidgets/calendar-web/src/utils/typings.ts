import { ObjectItem } from "mendix";
import { CalendarProps as ReactCalendarProps } from "react-big-calendar";
import { withDragAndDropProps } from "react-big-calendar/lib/addons/dragAndDrop";

export interface CalendarEvent {
    title: string;
    start: Date;
    end: Date;
    allDay: boolean;
    color?: string;
    item: ObjectItem;
}

export type EventDropOrResize = {
    event: CalendarEvent;
    start: Date;
    end: Date;
};

export interface DragAndDropCalendarProps<TEvent extends object = Event, TResource extends object = object>
    extends ReactCalendarProps<TEvent, TResource>,
        withDragAndDropProps<TEvent, TResource> {}
