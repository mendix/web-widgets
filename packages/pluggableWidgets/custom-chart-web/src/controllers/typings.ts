import { ActionValue, EditableValue } from "mendix";

export interface ControllerProps {
    configurationOptions: string;
    dataAttribute?: EditableValue<string>;
    dataStatic: string;
    eventDataAttribute?: EditableValue<string>;
    layoutAttribute?: EditableValue<string>;
    layoutStatic: string;
    onClick?: ActionValue;
    sampleData: string;
    sampleLayout: string;
}
