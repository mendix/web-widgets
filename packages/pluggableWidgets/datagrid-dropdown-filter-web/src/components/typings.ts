import { RefFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/RefFilterStore";
import { StaticSelectFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/StaticSelectFilterStore";

export interface EnumFilterAPI {
    filterStore: StaticSelectFilterStore;
    parentChannelName?: string;
}

export interface RefFilterAPI {
    filterStore: RefFilterStore;
    parentChannelName?: string;
}
