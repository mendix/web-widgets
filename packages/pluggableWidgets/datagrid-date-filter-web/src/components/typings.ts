import { Date_InputFilterInterface } from "@mendix/widget-plugin-filtering/typings/InputFilterInterface";

export interface DateFilterProps {
    filterStore: Date_InputFilterInterface;
    parentChannelName?: string;
}
