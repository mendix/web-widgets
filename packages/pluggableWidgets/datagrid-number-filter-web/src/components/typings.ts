import { ChangeEventHandler, CSSProperties } from "react";
import { FilterType } from "../../typings/FilterType";
import { Big } from "big.js";

interface FilterProps {
    adjustable: boolean;
    defaultFilterType: FilterType;
    className?: string;
    id?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    styles?: CSSProperties;
}

export interface FilterInputProps extends FilterProps {
    onFilterTypeClick: (type: FilterType) => void;
    onInputChange: ChangeEventHandler<HTMLInputElement>;
    inputValue: string;
    inputRef?: React.ClassAttributes<HTMLInputElement>["ref"];
    inputDisabled?: boolean;
}

export interface FilterComponentProps extends FilterProps {
    inputChangeDelay: number;
    value?: Big;
    updateFilters?: (value: Big | undefined, type: FilterType) => void;
    name: string;
    datagridChannelName?: string;
}
