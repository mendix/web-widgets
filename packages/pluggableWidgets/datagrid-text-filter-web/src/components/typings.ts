import { String_InputFilterInterface } from "@mendix/widget-plugin-filtering/typings/InputFilterInterface";

export interface StringFilterProps {
    filterStore: String_InputFilterInterface;
    parentChannelName?: string;
}
