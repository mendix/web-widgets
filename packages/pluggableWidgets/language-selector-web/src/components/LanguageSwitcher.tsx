import { FloatingFocusManager } from "@floating-ui/react";
import classNames from "classnames";
import { CSSProperties, ReactElement, useState } from "react";
import { PositionEnum, TriggerEnum } from "../../typings/LanguageSelectorProps";
import { useFloatingUI } from "../hooks/useFloatingUI";
import { LanguageItem } from "../LanguageSelector";

export interface LanguageSwitcherProps {
    className: string;
    currentLanguage: LanguageItem | undefined;
    languageList: LanguageItem[];
    onSelect?: (lang: LanguageItem) => void;
    position: PositionEnum;
    screenReaderLabelCaption?: string;
    style?: CSSProperties;
    tabIndex: number;
    trigger: TriggerEnum;
}

export const LanguageSwitcher = ({
    className,
    currentLanguage,
    languageList,
    onSelect,
    position,
    screenReaderLabelCaption,
    style,
    tabIndex,
    trigger
}: LanguageSwitcherProps): ReactElement => {
    const [isOpen, setOpen] = useState(false);

    const {
        activeIndex,
        context,
        floatingStyles,
        getFloatingProps,
        getItemProps,
        getReferenceProps,
        handleSelect,
        isTypingRef,
        listRef,
        refs
    } = useFloatingUI({
        currentLanguage,
        isOpen,
        languageList,
        onSelect,
        position,
        setOpen,
        triggerOn: trigger
    });

    return (
        <div className={classNames(className, "widget-language-selector", "popupmenu")} style={style}>
            <div
                ref={refs?.setReference}
                className={"popupmenu-trigger popupmenu-trigger-alignment"}
                aria-label={screenReaderLabelCaption || "Language selector"}
                aria-autocomplete="none"
                aria-activedescendant={isOpen && activeIndex !== null ? "" : undefined}
                tabIndex={tabIndex}
                {...getReferenceProps?.()}
            >
                <span className="current-language-text">{currentLanguage?.value || ""}</span>
                <span className="language-arrow" aria-hidden="true">
                    <div className={`arrow-image ${isOpen ? "arrow-up" : "arrow-down"}`} />
                </span>
            </div>
            {isOpen && (
                <FloatingFocusManager context={context!} modal={false}>
                    <div
                        aria-activedescendant={isOpen && activeIndex !== null ? "" : undefined}
                        className="popupmenu-menu"
                        ref={refs?.setFloating}
                        style={{
                            ...floatingStyles,
                            outline: 0,
                            overflowY: "auto"
                        }}
                        {...getFloatingProps?.()}
                    >
                        {languageList.map((item, index) => (
                            <div
                                key={item._guid}
                                ref={node => {
                                    listRef.current[index] = node;
                                }}
                                role="option"
                                tabIndex={index === activeIndex ? 0 : -1}
                                className={classNames("popupmenu-basic-item", {
                                    active: currentLanguage === item,
                                    highlighted: activeIndex === index
                                })}
                                {...getItemProps?.({
                                    onKeyDown(event) {
                                        if (event.key === "Enter") {
                                            event.preventDefault();
                                            handleSelect(index);
                                        }

                                        if (event.key === " " && !isTypingRef.current) {
                                            event.preventDefault();
                                            handleSelect(index);
                                        }
                                    },
                                    onClick() {
                                        handleSelect(index);
                                    }
                                })}
                            >
                                {item.value}
                            </div>
                        ))}
                    </div>
                </FloatingFocusManager>
            )}
        </div>
    );
};
