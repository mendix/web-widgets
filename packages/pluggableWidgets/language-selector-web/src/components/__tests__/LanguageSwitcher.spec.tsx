import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { PositionEnum, TriggerEnum } from "typings/LanguageSelectorProps";
import { LanguageSwitcher, LanguageSwitcherProps } from "../LanguageSwitcher";
import "@testing-library/jest-dom";

jest.useFakeTimers();

let props: LanguageSwitcherProps = {
    currentLanguage: undefined,
    languageList: [],
    position: "left" as PositionEnum,
    onSelect: jest.fn(),
    trigger: "click" as TriggerEnum,
    className: "",
    tabIndex: 0
};
const language = { _guid: "111", value: "En us" };

describe("Language switcher", () => {
    it("renders the structure with empty language list", async () => {
        const { asFragment } = render(<LanguageSwitcher {...props} />);
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
        const triggerElement = screen.getByRole("combobox");

        await user.click(triggerElement);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders the structure with language list and selected default language", async () => {
        props = { ...props, languageList: [language], currentLanguage: language };
        const { asFragment } = render(<LanguageSwitcher {...props} />);
        const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
        const triggerElement = screen.getByRole("combobox");

        await user.click(triggerElement);
        expect(asFragment()).toMatchSnapshot();
    });
});
