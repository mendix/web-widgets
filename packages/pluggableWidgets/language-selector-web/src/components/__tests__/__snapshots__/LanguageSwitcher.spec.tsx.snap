// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language switcher renders the structure with empty language list 1`] = `
<DocumentFragment>
  <div
    class="widget-language-selector popupmenu"
  >
    <div
      aria-activedescendant=""
      aria-autocomplete="none"
      aria-controls=":r0:"
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-label="Language selector"
      class="popupmenu-trigger popupmenu-trigger-alignment"
      data-floating-ui-inert=""
      role="combobox"
      tabindex="0"
    >
      <span
        class="current-language-text"
      />
      <span
        aria-hidden="true"
        class="language-arrow"
      >
        <div
          class="arrow-image arrow-up"
        />
      </span>
    </div>
    <div
      aria-activedescendant=""
      aria-orientation="vertical"
      class="popupmenu-menu"
      data-floating-ui-focusable=""
      id=":r0:"
      role="listbox"
      style="position: fixed; left: 0px; top: 0px; outline: 0; overflow-y: auto; transform: translate(0px, 0px);"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`Language switcher renders the structure with language list and selected default language 1`] = `
<DocumentFragment>
  <div
    class="widget-language-selector popupmenu"
  >
    <div
      aria-activedescendant=""
      aria-autocomplete="none"
      aria-controls=":r2:"
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-label="Language selector"
      class="popupmenu-trigger popupmenu-trigger-alignment"
      data-floating-ui-inert=""
      role="combobox"
      tabindex="0"
    >
      <span
        class="current-language-text"
      >
        En us
      </span>
      <span
        aria-hidden="true"
        class="language-arrow"
      >
        <div
          class="arrow-image arrow-up"
        />
      </span>
    </div>
    <div
      aria-activedescendant=""
      aria-orientation="vertical"
      class="popupmenu-menu"
      data-floating-ui-focusable=""
      id=":r2:"
      role="listbox"
      style="position: fixed; left: 0px; top: 0px; outline: 0; overflow-y: auto; transform: translate(0px, 0px);"
      tabindex="-1"
    >
      <div
        class="popupmenu-basic-item active highlighted"
        role="option"
        tabindex="0"
      >
        En us
      </div>
    </div>
  </div>
</DocumentFragment>
`;
